cmd/apiserver/main.go

```diff
@@ -156,7 +156,18 @@ func main() {
 			ClientAuth:         tls.VerifyClientCertIfGiven,
 			InsecureSkipVerify: tlsNoVerify,
 			ClientCAs:          tlsTrustedCAs,
-			MinVersion:         tls.VersionTLS11,
+			// TLS < 1.1 is deprecated
+			MinVersion:         tls.VersionTLS12,
+			// Limit cipher suite options to those currently recommended (e.g. avoid CBC,
+			// ensure ECDHe for Forward Secrecy).
```

Comment:
Can you please cite where these recommendations are coming from?

Reply:
https://ciphersuite.info/cs/ is a good resource in general although the list itself is way beyond those supported/relevant to the `go` TLS library.

Where would you like this citation? In the code comments or the commit / MR message? I'm also happy to propose an update to the docs if appropriate, but it's more wider-reaching than just `apiserver`.

Reply:
I appreciate the resource list, it is certainly helpful and could be a part of the commit message.

I am looking for something around accepted standards, e.g. in NIST guidelines.

It does seem the ciphers have [FIPS 140 compliance](https://wiki.openssl.org/index.php/FIPS_mode_and_TLS#TLS_1.2) (and that would be a better URL to point to within the codebase - I did a check using `openssl ciphers -tls1_2 -v 'TLSv1.2+FIPS:!eNULL:!aNULL'` on a EL8 build of OpenSSL) EXCEPT for the CHACHA20. However, if we have several FIPS 140 compliant ciphers available, that should suffice.

Reply:
That's a reasonable baseline to reference, yes.

I also test endpoints using [SSLLabs](https://www.ssllabs.com/ssltest/), or the containerised equivalent, [`ssltest.sh`](https://hub.docker.com/r/drwetter/testssl.sh). SSLLabs has a [best practices guide](https://github.com/ssllabs/research/wiki/SSL-and-TLS-Deployment-Best-Practices), which might be a better resource than FIPS 140. That said, the document has some contradictions with their own test engine, as some of the ciphers they recommend there are now [discouraged in their tests for confidence reasons](https://blog.qualys.com/product-tech/2019/04/22/zombie-poodle-and-goldendoodle-vulnerabilities):

> SSL Labs identifies cipher suites using CBC with orange color and with text WEAK. This change won’t have any effect on the grades, as it only means that SSL Labs discourages the use of CBC-based cipher suites further.

I'll look at updating the default cipher list based on the FIPS 140/SSL Labs lists. Leaving ChaCha out is fine; it's a faster cipher on certain mobile devices, which isn't really relevant for PGO, and is offered in TLSv1.3 anyway, should it become desirable as time progresses,

Reply:
@jkatz I've updated the PR with three references that, in combination with what's available in Go TLS, justify the cipher list that's in there.

Are there any API tests that I could refer to to ensure all is well?

Documentation-wise, I would say the code comments are adequate for now, along with a mention of the minimum TLS version in whatever release this makes it into.

If a future feature enhancement comes along that makes these user-configurable, then those comments can be hauled up into the documentation and used to justify the defaults and give users material to refer to should they want to make their own changes.

Reply:
> I left ChaCha in because although it's not in the FIPS list, it's in SSLLab's recommendations to use AEAD ciphers.
:+1: That checks out in my book.

> Are there any API tests that I could refer to to ensure all is well?

We have an e2e testing suite, though I don't believe it exercises all the TLS bits at the moment: https://github.com/CrunchyData/postgres-operator/blob/master/testing/pgo_cli/README.md

> Documentation-wise, I would say the code comments are adequate for now, along with a mention of the minimum TLS version in whatever release this makes it into.

I'll review and judge that :wink: But yes, I agree, it's much better cited. Thanks!

> If a future feature enhancement comes along that makes these user-configurable, then those comments can be hauled up into the documentation and used to justify the defaults and give users material to refer to should they want to make their own changes.

Agreed. Thanks!

Reply:
> We have an e2e testing suite, though I don't believe it exercises all the TLS bits at the moment: https://github.com/CrunchyData/postgres-operator/blob/master/testing/pgo_cli/README.md

Thanks. I'm having a look at that, as it's not run in CI, to see what I can recreate locally. I haven't done any self-built deployments of PGO yet so it may take me a while to get some time to have a go at it, although it seems clear enough.